src/v/cloud_storage/topic_recovery_service.cc

```diff
@@ -0,0 +1,594 @@
+                      "topic recovery completed with error: {}",
+                      result.error().context);
+                } else {
+                    vlog(
+                      cst_log.info,
+                      "topic recovery completed. Pending downloads");
+                }
+            });
+        });
+        return {
+          .status_code = ss::httpd::reply::status_type::accepted,
+          .message = "recovery started"};
+    } catch (const bad_request& ex) {
+        return {
+          .status_code = ss::httpd::reply::status_type::bad_request,
+          .message = fmt::format(
+            "bad recovery request payload: {}", ex.what())};
+    } catch (const std::exception& ex) {
+        return {
+          .status_code = ss::httpd::reply::status_type::internal_server_error,
+          .message = fmt::format(
+            "recovery init failed with error: {}", ex.what())};
+    }
+}
+
+ss::future<> topic_recovery_service::propagate_state(state s) {
+    return container().invoke_on_all([s](auto& svc) {
+        svc._state = s;
+        return ss::make_ready_future<>();
+    });
+}
+
+ss::future<result<void, recovery_error_ctx>>
+topic_recovery_service::start_bg_recovery_task(recovery_request request) {
+    if (is_active()) {
+        vlog(cst_log.warn, "A recovery is already active");
+        co_return recovery_error_ctx::make(
+          "A recovery is already active",
+          recovery_error_code::recovery_already_running);
+    }
+
+    vlog(cst_log.info, "Starting recovery task with request: {}", request);
+    // Setting this state here ensures that another request coming in soon after
+    // on the same shard is rejected. If we set this state after the RPC call,
+    // it is possible that two requests sent back to back on the controller
+    // leader assume that no other recovery is running.
+    _state = state::starting;
+    if (_pending_status_timer.cancel()) {
+        vlog(
+          cst_log.warn,
+          "stopped running pending status timer before starting recovery "
+          "request");
+    }
+
+    vlog(
+      cst_log.debug, "checking if recovery is already running on another node");
+
+    // When making RPC call we make sure that we do not check the state of the
+    // current shard, because recovery is in starting state here.
+    if (co_await _topic_recovery_status_frontend.local().is_recovery_running(
+          container(),
+          cluster::topic_recovery_status_frontend::skip_this_node::yes)) {
+        vlog(cst_log.warn, "A recovery is already active");
+        _state = state::inactive;
+        co_return recovery_error_ctx::make(
+          "A recovery is already active",
+          recovery_error_code::recovery_already_running);
+    }
+
+    auto fib = make_rtc(_as, _config);
+
+    co_await propagate_state(state::scanning_bucket);
+    vlog(cst_log.debug, "scanning bucket {}", _config.bucket);
+    auto bucket_contents_result = co_await _remote.local().list_objects(
```

Comment:
It makes sense to list selectively, that would be a relatively significant change and I think it will require adding more tests. Does it make sense to do it in a follow up PR after this one? 

WRT. resources usage, I also think we can farm out the downloading of manifests and creating topics to other shards on the same node to speed up the recovery, as holding too many manifests in memory can be a problem and the problem can be parallelized easily. The tricky part would be resetting the topics by watching the bucket for result files.

Reply:
Yes, this can go in a followup, just needs to be in before we release.

Reply:
Right now this value is the `recovery_task_config::operation_timeout_ms` which is read from `config::shard_local_cfg().cloud_storage_manifest_upload_timeout_ms.value`, it is used for all operations which is not ideal. 

I can add a multiplier for the `list_objects` operation and look into introducing properties for this in a follow up.

Reply:
I have added a multiplier of 10 for the list call. The default operation timeout is 10 seconds. Will look into adding a property for this.